### ISSUE TITLE
[feature request] Multi-layer bidirectional RNN

### ISSUE THREAD

**Issue by adonisues** [pytorch/pytorch#4930]
Hello 
I request another type of multi-layer bidirectional RNN.
Currently forward output and backward output is concatenated after each layer.
But, for language modeling, we need independent forward rnn and backward rnn util the last layer and output concatenation is only need at the last layer.

thank you.

**Comment by SsnL** [comment]
A couple days ago, @csarofeen pointed out that the non-cudnn and cudnn versions of bidirectional RNN behave differently. With cudnn, forward and backward directions are linearly independent. So if you are certain that you are using cudnn, it will work fine for you. That said, this is an inconsistency that we should fix.

**Comment by adonisues** [comment]
Thank you for your fast reply.
I draw two different types multi-layer bidirectional LSTM. 
I think the type 2 in the figure is supported now using " nn.LSTM(nlayers=2, bidirectional=True)"
My request is the type 1 model which is totally independent forward and backward layers.

Thank you
 

![2kinds_of_multilayer_bidirectional_lstm](https://user-images.githubusercontent.com/14052287/35610748-dd41a23a-06a5-11e8-803e-c577359c372b.jpg)

**Comment by Stonesjtu** [comment]
Seems like the type 1 adds a uni-directional LSTM with reverse training data.

**Comment by SsnL** [comment]
@adonisues Yes, your plots are exactly what I meant about the difference between the cudnn and non cudnn LSTMs. When using cudnn, `nn.LSTM` is type 1. When it is not using cudnn, it is type 2.

**Comment by adonisues** [comment]
Hello 
Thank you for your kind reply @SsnL.
I did some experiment  on my server. Code and output is below

In my code, the LSTM looks like Type 2 because the second layer has double size input (6) .
Input size of the first layer is 3.
As your reply, i am using non-cudnn LSTM now.
Can you show me the way using Cudnn LSTM ? 
I can't find any special option except  "torch.backends.cudnn.enabled = True"
 
Thank you

pytorch version is 0.3.0.post4

Code -------------------------------------
import torch.backends.cudnn as cudnn
import torch.backends.cudnn.rnn
import torch.nn as nn

torch.backends.cudnn.enabled = True

print "torch.backends.cudnn.enabled : ", torch.backends.cudnn.enabled
print "torch.backends.cudnn.version : ", torch.backends.cudnn.version()
print "torch.backends.cudnn.is_acceptable : ", torch.backends.cudnn.is_acceptable(torch.cuda.FloatTensor(1))

model1 = nn.LSTM(2, 3, 2, bidirectional=True)
print model1
for parameter in model1.parameters() :
    print (parameter.size())
---------------------------------------------------------------------
Output------------------------------------------------------------ 
torch.backends.cudnn.enabled :  True
torch.backends.cudnn.version :  7003
torch.backends.cudnn.is_acceptable :  True
LSTM(2, 3, num_layers=2, bidirectional=True)
(12L, 2L)
(12L, 3L)
(12L,)
(12L,)
(12L, 2L)
(12L, 3L)
(12L,)
(12L,)
**(12L, 6L)**
(12L, 3L)
(12L,)
(12L,)
**(12L, 6L)**
(12L, 3L)
(12L,)
(12L,)
--------------------------------------------------------------

**Comment by SsnL** [comment]
@adonisues Yes, I am aware of the weight shapes. The doubled weight sizes look like type 2, but it is actually just type 1 when using cudnn. It is difficult to generally determine whether the lstm is using cudnn path or not, but fortunately we see from here(https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/rnn.py#L300) that it just needs to know if the input is cudnn_acceptable or not. However, be careful that this condition is subject to change in future versions.

**Comment by adonisues** [comment]
I checked "cudnn.is_acceptable(input.data)" is True when torch.backends.cudnn.enabled is True.
I am using Cudnn LSTM now.

Unfortunately, I can't find any evidence Cudnn LSTM is Type1. 
But model parameters and my code for Language Model show that Cudnn LSTM and non-Cudnn LSTM are type 2. 
If Cudnn LSTM is type 1, there are redundant parameters. This is very strange.   

Can you show me another manuscript or wed document can support  the Cudnn LSTM is Type 1 ?

**Comment by SsnL** [comment]
> But model parameters and my code for Language Model show that Cudnn LSTM and non-Cudnn LSTM are type 2.

What I was trying to say is that do not look at the model parameter shapes as they are deceiving. All the cudnn wrapper does is to give a contiguous chunk of GPU memory to cudnn. There are not redundant parameters, it's just PyTorch gives cudnn more than what it needs.

doc:
http://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnDirectionMode_t

**Comment by adonisues** [comment]
@SsnL 
I checked cudnn LSTM and non-cudnn LSTM as below
When torch.backends.cudnn.enabled is True and False, results are same.
If non-cudnn and cudnn lstm are different, results should be different. But it is not.
Result is exactly same.

'''
import torch
import torch.backends.cudnn as cudnn
import torch.nn as nn
from torch.autograd import Variable
import torch.nn.init as init

torch.backends.cudnn.enabled = True

print "torch.backends.cudnn.enabled : ", torch.backends.cudnn.enabled
print "torch.backends.cudnn.version : ", torch.backends.cudnn.version()

model = nn.LSTM(1, 2, 2, bidirectional=True).cuda()

for weights in  model.all_weights :
    for weight in weights :
        init.constant(weight, 0.1)

input = Variable(torch.FloatTensor([[[1],[0.1]],[[2],[0.2]]]).cuda())
output, hidden = model(input)

print "cudnn.is_acceptable(input.data) : ", cudnn.is_acceptable(input.data)
print output
'''
ouput
'''
(0 ,.,.) = 
  0.0786  0.0786  0.1302  0.1302
  0.0710  0.0710  0.1154  0.1154

(1 ,.,.) = 
  0.1324  0.1324  0.0828  0.0828
  0.1156  0.1156  0.0713  0.0713
'''

**Comment by adonisues** [comment]
@SsnL 
Your reference http://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnDirectionMode_t said,
Cudnn LSTM is type 2 as below.

**CUDNN_BIDIRECTIONAL**
Each layer of the the network iterates recurrently from the first input to the last and separately from the last input to the first. The outputs of the two are concatenated at each iteration giving the output of the layer.

Thank you.

**Comment by SsnL** [comment]
@adonisues You made some very good results and valid points. This proved that my previous understanding was likely wrong. @csarofeen Could it be that autograd and cudnn multilayer bidierctional RNNs are actually the same?

**Comment by csarofeen** [comment]
@adonisues @ssnl cudnn is used by default. So unless you use torch.backends.cudnn.enabled = False you will be using it for any situation that matches cuDNN features. This can be shown in the annoying cases where people with attention try to use cuDNN RNN step by step and they have to turn it off because it isn't efficient with the way it's implemented in PyTorch.

I talked with the Jeremy Appleyard after I had difficulty validating the version of PyTorch RNN's I'm working on and he confirmed cuDNN is type 1. PyTorch is actually a mixture of 1 and 2. It has weights the size of 1 but runs it like 2.

I have a new RNN implementation I've been working on for PyTorch but tests aren't finished, it's not feature complete, and I don't have the fallback to cuDNN yet. I can post an early version if you guys want to help with it.

**Comment by csarofeen** [comment]
@adonisues @ssnl I was sure it was wrong, but looks like I was wrong. https://gist.github.com/csarofeen/3ebb607d2a7c2c1093a134e356b9e867
The only error I can see is batch_first doesn't match cudnn/pytorch. Looks like I will have to re-read what pytorch is doing.

**Comment by csarofeen** [comment]
The thing that doesn't make sense to me is: https://gist.github.com/csarofeen/141bb62c49f1d1b8e867a8d50ea62a4e
prints:
```
Seq_size 2
hidden_size 3
inp_size 4
batch_size 5
num_layers 2
weight_ih_l0 torch.Size([12, 4])
weight_hh_l0 torch.Size([12, 3])
weight_ih_l0_reverse torch.Size([12, 4])
weight_hh_l0_reverse torch.Size([12, 3])
weight_ih_l1 torch.Size([12, 6])
weight_hh_l1 torch.Size([12, 3])
weight_ih_l1_reverse torch.Size([12, 6])
weight_hh_l1_reverse torch.Size([12, 3])
```
Here the ih_l1 weights don't make sense to me, shouldn't they be same size as hh?

**Comment by adonisues** [comment]
@csarofeen @SsnL 
That is why cudnn and non-cudnn LSTM is Type 2 in my figure. 

Outputs of forward and backward layer are concatenated after each bidirectional layer.
When the first forward layer output is 3 and backward output is 3, the input vector of the second layer is 6. 

This type 2 LSTM is usually used for classification work.   
But it doesn't work for language modeling because the label of input word should be unknown.

I think if you support type 1 LSTM, many researcher in language can be happy ^^

Thank you.

**Comment by Stonesjtu** [comment]
@adonisues  As noted in your figure, you can setup an LSTM with `<s> w1 w2 </s>`, and another LSTM with `</s> w2 w1 <s>`, simply concatenating their outputs is enough for a *type 1* bi-directional LSTM.

@csarofeen  the output of `l0` consists of the outputs of both directions (3 * 2) in `l0`, and `12 = 4(gates) * 3(hidden)`,  then `(12, 6)` makes sense. At the same time the hidden states of two directions in `l1` are independent so we have `(12, 3)` .

**Comment by Stonesjtu** [comment]
@csarofeen As there are 2 `(12, 4)` in the first layer, each for one direction. the input size of the `l0` is `4`, while the input size of `l1` is `3(hidden of l0) * 2 (2 directions in l0)`. 

`input`     `l0`     `output of l0` 
4 -----> 3 ------>6(combined)
-------> 3 ------->

**Comment by csarofeen** [comment]
@Stonesjtu 
I am well aware of the first layer.
There's a weight for forward and backward with l1 labels. Both of sizes that could handle forward and backward together. There should be one, not the other, or they should be half the size in the second dimension like hidden, no?
weight_ih_l1 torch.Size([12, 6])
weight_ih_l1_reverse torch.Size([12, 6])

**Comment by jekbradbury** [comment]
In each layer, the forward and backward RNNs each take input from both the forward and backward hidden states of the previous layer. Isn't that also cuDNN's behavior?

**Comment by csarofeen** [comment]
What I was told about cuDNN that I think needs to be further verified, is yes, HOWEVER the multiplication of the weights are done so they might as well be separate inputs. In the kernel it is just indexing math to align the right portion of the weight to the right multiplication. Since they seem to match with the first gist I put in this issue. I really think both implementations need to be gone through in detail to figure out how they match and how it is being done.

**Comment by erikvdplas** [comment]
@adonisues did you ever find a good way to get the desired ('type 1') behavior?
Currently experimenting with two separate multi-layer RNNs, where one gets its input in reversed order, then combining their outputs. Will keep you posted if that works.

**Comment by sailordiary** [comment]
> @adonisues did you ever find a good way to get the desired ('type 1') behavior?
> Currently experimenting with two separate multi-layer RNNs, where one gets its input in reversed order, then combining their outputs. Will keep you posted if that works.

Hi @erikvdplas -- any updates on this?

**Comment by erikvdplas** [comment]
Yes @sailordiary! Sorry for not following-up on this thread. My implementation [here](https://github.com/erikvdplas/binding-site-predictor/blob/master/model.py) (with the gist of it being in line 50-53) seemed to work as desired. I think the code is pretty straightforward. Let me know if something's not clear.

**Comment by kishorepv** [comment]
> Thank you for your fast reply.
> I draw two different types multi-layer bidirectional LSTM.
> I think the type 2 in the figure is supported now using " nn.LSTM(nlayers=2, bidirectional=True)"
> My request is the type 1 model which is totally independent forward and backward layers.
> 
> Thank you
> 
> ![2kinds_of_multilayer_bidirectional_lstm](https://user-images.githubusercontent.com/14052287/35610748-dd41a23a-06a5-11e8-803e-c577359c372b.jpg)

Which is the type we have in the current version PyTorch (1.x and above?)

**Comment by ezyang** [comment]
I'm not too sure about the status of this issue, but tentatively marking in high priority based on user activity

**Comment by andfoy** [comment]
@ezyang, is it necessary to verify that the model currently implemented corresponds to the Type-1? Or is it necessary to introduce a new flag in nn.LSTM for allowing both behaviours?

**Comment by andfoy** [comment]
@rgommers @ezyang  I've confirmed that both the CPU native and CUDNN implementations of bidirectional RNNs are of type 2. Here are some of the findings:

## CPU-bound RNN

The CPU-part can be easily adapted in order to add a flag to independent units for both backward and forward:

```c++
      // ATen/native/RNN.cpp: 570-590
      auto input_w = params.first.linear_ih(input);
      step_inputs = input_w.unbind(0);
      auto fw_result = layer_(
          step_inputs, input_hidden.first, params.first, true);
      auto fw_output = at::stack(fw_result.outputs, 0);
      input_w = params.second.linear_ih(input);
      step_inputs = input_w.unbind(0);
      auto rev_step_inputs = reverse(std::move(step_inputs));
      auto rev_result =
          layer_(rev_step_inputs, input_hidden.second, params.second, true);
      std::reverse(rev_result.outputs.begin(), rev_result.outputs.end());
      auto rev_output = at::stack(rev_result.outputs, 0);
      // at:cat is responsible for the Type-2 behaviour
      return {at::cat({fw_output, rev_output}, fw_output.dim() - 1),
              std::make_pair(fw_result.final_hidden, rev_result.final_hidden)};
``` 
## CUDNN implementation
As previous users have mentioned, the CUDNN flag ``CUDNN_BIDIRECTIONAL`` for the call ``cudnnRNNForwardTraining`` enforces a Type-2 implementation, as it is mentioned on the docs:

> **CUDNN_BIDIRECTIONAL**
    Each layer of the network iterates recurrently from the first input to the last and separately from the last input to the first. The outputs of the two are concatenated at each iteration giving the output of the layer. 

As the CUDNN implementation handles the layers implicitly, the only way to implement a Type-1 CUDNN RNN is introducing a loop in C++ to handle each layer, while CUDNN makes the computation for a single one:

```c++
// Example
for(int layer = 0; layer < layers; layer++)
{
    AT_CUDNN_CHECK(cudnnRNNForwardTraining(
          handle,
          descs.rnn_desc.desc(),
          fn.tensors.seq_length,
          x_descs_arr.data(), x.data_ptr(),
          descs.hx_desc.desc(), hx.data_ptr(),
          descs.cx_desc.desc(), cx.defined() ? cx.data_ptr() : nullptr,
          w_desc.desc(), weight_buf.data_ptr(),
          y_descs_arr.data(), y.data_ptr(),
          descs.hy_desc.desc(), hy.data_ptr(),
          descs.cy_desc.desc(), cy.defined() ? cy.data_ptr() : nullptr,
          workspace.data_ptr(), workspace.size(0),
          reserve.data_ptr(), reserve.size(0)
          ));
}
```

## MIOpen/ROC implementation
By taking a look at the MIOpen documentation regarding their RNN implementation, it would seem that the API is pretty similar to the CUDA one, thus, one would expect that the bidirectional behaviour is the same as CUDA, that is, Type 2. Once again, as MIOpen handles the layer iterations internally, the only solution would be to handle the iterations directly on PyTorch and call MIOpen on each one of them.

```c++
for(int layer = 0; layer < layers; layer++)
{
     MIOPEN_CHECK(miopenRNNForwardTraining(handle, descs.rnn_desc.desc(),
                fn.tensors.seq_length,
                x_descs_arr.data(), x.data_ptr(),
                descs.hx_desc.desc(), hx.data_ptr(),
                descs.cx_desc.desc(), cx.defined() ? cx.data_ptr() : nullptr,
                w_desc.desc(), weight_buf.data_ptr(),
                y_descs_arr.data(), y.data_ptr(),
                descs.hy_desc.desc(), hy.data_ptr(),
                descs.cy_desc.desc(), cy.defined() ? cy.data_ptr() : nullptr,
                workspace.data_ptr(), workspace_size, reserve.data_ptr(), reserver_size ));
}
```

While the CPU implementation could be done easily, I don't know what would be the performance impacts of handling layers directly on PyTorch as opposed of letting the GPU libraries handling the layers instead. What do you think about it?

**Comment by ezyang** [comment]
There will definitely be a performance loss for handling the layers separately in PyTorch, but since it's pretty easy to do we should go ahead and do it. Maybe one of folks at NVIDIA (@mcarilli perhaps) can check if cudnn secretly supports Type 1, or if this is the best we can do without writing our own kernel (out of scope for this issue, IMO).

**Comment by rgommers** [comment]
The PR for this is ready (gh-33794), if anyone who requested the functionality would like to test that PR, that'd be very welcome.
